packages/hash/frontend/src/pages/shared/ory-kratos.ts

```diff
@@ -0,0 +1,191 @@
+          }
+          return;
+        }
+        case "session_already_available":
+          // User is already signed in, let's redirect them home!
+          await router.push("/");
+          return;
+        case "session_refresh_required": {
+          // We need to re-authenticate to perform this action
+          const { redirect_browser_to } =
+            kratosError as NeedsPrivilegedSessionError;
+
+          if (redirect_browser_to) {
+            await router.replace(redirect_browser_to);
+          }
+          return;
+        }
+        case "self_service_flow_return_to_forbidden":
+          // The flow expired, let's request a new one.
+          setErrorMessage("The return_to address is not allowed.");
+          setFlow(undefined);
+          await router.push(`/${flowType}`);
+          return;
+        case "self_service_flow_expired":
+          // The flow expired, let's request a new one.
+          setErrorMessage(
+            "Your interaction expired, please fill out the form again.",
+          );
+          setFlow(undefined);
+          await router.push(`/${flowType}`);
+          return;
+        case "security_csrf_violation":
+          // A CSRF violation occurred. Best to just refresh the flow!
+          setErrorMessage(
+            "A security violation was detected, please fill out the form again.",
+          );
+          setFlow(undefined);
+          await router.push(`/${flowType}`);
+          return;
+        case "security_identity_mismatch":
+          // The requested item was intended for someone else. Let's request a new flow...
+          setFlow(undefined);
+          await router.push(`/${flowType}`);
+          return;
+        case "browser_location_change_required": {
+          // Ory Kratos asked us to point the user to this URL.
+          const { redirect_browser_to } =
+            kratosError as SelfServiceBrowserLocationChangeRequiredError;
+
+          if (redirect_browser_to) {
+            await router.replace(redirect_browser_to);
+          }
+          return;
+        }
+      }
+    }
+
+    switch (err.response?.status) {
+      case 410:
+        // The flow expired, let's request a new one.
+        setFlow(undefined);
+        await router.push(`/${flowType}`);
+        return;
+    }
+
+    // We are not able to handle the error? Return it.
+    return Promise.reject(err);
+  };
+
+const maybeGetCsrfTokenFromFlow = (flow: SelfServiceFlow) =>
+  flow.ui.nodes
+    .map(({ attributes }) => attributes)
+    .filter(isUiNodeInputAttributes)
+    .find(({ name }) => name === "csrf_token")?.value;
+
+export const mustGetCsrfTokenFromFlow = (flow: SelfServiceFlow): string => {
+  const csrf_token = maybeGetCsrfTokenFromFlow(flow);
+
+  if (!csrf_token) {
+    throw new Error("CSRF token not found in flow");
+  }
+
+  return csrf_token;
+};
+
+// Returns a function which will log the user out
+export const useLogoutFlow = (deps?: DependencyList) => {
```

Comment:
we will need to bring in the logic from the existing `useLogout` to reset the ApolloClient cache

Reply:
Good shout, I think this hook should replace the previous `useLogout` hook as the HASH workspace GQL API will no longer be used to log users out. I've replaced the hook, and added the missing `client.resetStore()` in https://github.com/hashintel/hash/pull/889/commits/8f90c07ce79c758340375d2852eec68d297b877b.

Currently calling `client.resetStore()` throws a 500 internal server error, but I think that's because it's trying to call the GQL `me` mutation and is encountering an internal server error because the `me` query is currently broken (it's trying to query a database that doesn't exist anymore).

Therefore using the logout button now gives us an error, but I think that should be fixed in the next-up PR.